### PR TITLE
New function to expose more info about an installed package

### PR DESCRIPTION
**Pull request by jennybc**
Addresses #1957

Requires https://github.com/posit-dev/ark/pull/625

What's the payoff?

* Sometimes you want to get info about a package version without necessarily challenging the user to install/upgrade. (I need this to finish #5231, for example). 
* If you do need to invite the user to install or upgrade a package, this extra info allows us to build a better message.

### QA Notes

To experience that `checkInstalled()` still works and to see the user-facing message in the absence of a minimum version requirement, here's one idea:

* Remove the styler package: `remove.packages("styler")`
* Trigger the *Format Document* command

You should see this:

Screenshot 2024-11-13 at 3 46 39 PM

To experience that `checkInstalled()` still works and to see the user-facing message in the presence of a minimum version requirement, here's one idea:

* Install renv at a version below positron-r's minimum version, which is currently 1.0.9. I would do this with `pak::pak("renv@1.0.8")`.
* Walk through *File > New Project > R Project* and click the box to use renv:

  Screenshot 2024-11-13 at 3 35 45 PM

As the new project is being stood up, you should see this:

Screenshot 2024-11-13 at 3 37 12 PM

### CODE REVIEW

**Review by juliasilge — COMMENTED** [review]
I tend to think we want to keep the per-session cache of what we have already looked up, given that this goes all the way through to the R session and we call this every time we install, check, document, etc.

Were you thinking differently about this?

**Comment by jennybc** [comment]
I'm going to bring caching back with a bit of modification. Stay tuned.

File: extensions/positron-r/src/session.ts [hunk 1]
```diff
@@ -385,71 +388,155 @@ export class RSession implements positron.LanguageRuntimeSession, vscode.Disposa
...
+			id,
+			positron.RuntimeCodeExecutionMode.Interactive,
+			positron.RuntimeErrorBehavior.Continue);
+
+		// Wait for the the runtime to be idle, or for the timeout:
+		await Promise.race([promise, timeout(2e4, 'waiting for package installation')]);
+
+		pkgInst = await this.packageVersion(pkgName, minimumVersion);
+		compatible = pkgInst?.compatible ?? false;
+		return compatible;
```
Comment (jennybc) [inline extensions/positron-r/src/session.ts:539]: This is a bit new: previously we always assumed that our attempt to install/update the package was successful. Now we check that specifically, cache new version info, and return the result (hopefully `true`).

**Comment by jennybc** [comment]
OK @juliasilge this is done, now, complete with restored and upgraded caching. Additional work that is not shown here: I temporarily added lots of logging to verify that the cache was being used/updated as expected and I was satisfied with the results.

File: extensions/positron-r/src/session.ts [hunk 2]
```diff
@@ -33,6 +35,7 @@ interface EnvVar {
 // locale to also be present here, such as LC_CTYPE or LC_TIME. These can vary by OS, so this
 // interface doesn't attempt to enumerate them.
 interface Locale {
+	// eslint-disable-next-line @typescript-eslint/naming-convention
```
Comment (juliasilge) [inline extensions/positron-r/src/session.ts:38]: Thank you so much 😭

**Review by juliasilge — CHANGES_REQUESTED** [review]
I think something here isn't quite working right yet.

With our current release build, if I try to create a new project using renv but I have renv 1.0.8 installed, I see this:

Screenshot 2024-11-15 at 6 10 36 PM

With a dev build in the same situation, the new project opens up but there is no message or prompt to install renv.

Maybe it's just that ark needs to be bumped in this PR?

**Comment by jennybc** [comment]
I have already rebased and force pushed this PR branch to include the ark version bump I did earlier today. Do you need to refetch this branch? Any chance you've got a local ark build that's getting used in the dev build instead of the 0.1.154 release from this afternoon?

**Comment by jennybc** [comment]
I just re-did the QA step described above and I'm still getting the desired pop-up:

Screenshot 2024-11-15 at 5 39 58 PM

**Comment by juliasilge** [comment]
I think it's possible what we need is step 4 [here](https://connect.posit.it/positron-wiki/updating-amalthea.html#pushing-kernel-updates), like [this commit](https://github.com/posit-dev/positron/pull/4606/commits/25eb1eb6829f5ed6e9f83b745cd4c6dc8ea29c86).

**Comment by jennybc** [comment]
🤔 step 4 has happened, in `main`, in 34695a1276d865021bc4ffaac5a0746615956341. And this PR is rebased on that.

Here's the ark requirement in `positron/extensions/positron-r/package.json` in this branch:

https://github.com/posit-dev/positron/blob/d0e363d222ac0d6936be3925281902a5be82a2ee/extensions/positron-r/package.json#L673

It does feel like an ark mismatch. Does `.ps.ark.version()` report 0.1.154 in your dev build?

**Comment by jennybc** [comment]
I made one more change, but it would not explain what you're seeing (a complete lack of prompt).

**Review by juliasilge — APPROVED** [review]
I did a full `rebuild.sh` and now see the correct prompt:

Screenshot 2024-11-17 at 4 59 07 PM

I'm not totally sure what the problem was (no local build of ark, etc) but thank you for your patience!

**Pull request merged**